designs/dragonphy_top/constraints/gen_constraints.py

```diff
@@ -32,19 +33,83 @@
     -waveform {{0 {0.5*main_per*time_scale}}} \\
     [get_pins iacore/clk_adc]
 
-# TX clocks
-# TODO: add generated clocks in TX
+#############
+# TX clocks #
+#############
 
-create_clock -name clk_prbsgen \\
-    -period {main_per*time_scale} \\
-    -waveform {{0 {0.5*main_per*time_scale}}} \\
-    [get_pins itx/clk_prbsgen]
+# PI outputs
+
+create_clock -name clk_tx_pi_0 \\
+    -period {clk_4x_per} \\
+    -waveform {{0 {0.5*clk_4x_per}}} \\
+    [get_pins itx/iPI[0].iPI/clk_out_slice]
+
+create_clock -name clk_tx_pi_1 \\
+    -period {clk_4x_per} \\
+    -waveform {{{0.25*clk_4x_per} {0.75*clk_4x_per}}} \\
+    [get_pins itx/iPI[1].iPI/clk_out_slice]
+
+create_clock -name clk_tx_pi_2 \\
+    -period {clk_4x_per} \\
+    -waveform {{{0.5*clk_4x_per} {clk_4x_per}}} \\
+    [get_pins itx/iPI[2].iPI/clk_out_slice]
+
+create_clock -name clk_tx_pi_3 \\
+    -period {clk_4x_per} \\
+    -waveform {{{0.75*clk_4x_per} {1.25*clk_4x_per}}} \\
+    [get_pins itx/iPI[3].iPI/clk_out_slice]
 
-# clock uncertainty
+# clock dividers
+
+create_generated_clock -name clk_tx_hr_0 \\
+    -source [get_pins itx/qr_mux_4t1_0/div/clkin] \\
+    -divide_by 2 \\
+    [get_pins itx/qr_mux_4t1_0/div/clkout]
+
+create_generated_clock -name clk_tx_hr_1 \\
+    -source [get_pins itx/qr_mux_4t1_1/div/clkin] \\
+    -divide_by 2 \\
+    [get_pins itx/qr_mux_4t1_1/div/clkout]
+
+create_generated_clock -name clk_tx_qr_0 \\
+    -source [get_pins itx/hr_mux_16t4_0/clk_div/clkin] \\
+    -divide_by 2 \\
+    [get_pins itx/hr_mux_16t4_0/clk_div/clkout]
+
+create_generated_clock -name clk_tx_qr_1 \\
+    -source [get_pins itx/hr_mux_16t4_1/clk_div/clkin] \\
+    -divide_by 2 \\
+    [get_pins itx/hr_mux_16t4_1/clk_div/clkout]
+
+#####################
+# clock uncertainty #
+#####################
+
+# clk_retimer
 set_clock_uncertainty -setup 0.03 clk_retimer
 set_clock_uncertainty -hold 0.03 clk_retimer
-set_clock_uncertainty -setup 0.03 clk_prbsgen
-set_clock_uncertainty -hold 0.03 clk_prbsgen
+
```

Comment:
Clock uncertainties set to a few percent of the clock period.

Reply:
Does this `uncertainty` specify the clock jitter?
What is the value stands for, rms/pk-pk?

thanks for adding this.

> Clock uncertainties set to a few percent of the clock period.